packages/backend/server/src/plugins/oauth/providers/oidc.ts

```diff
@@ -37,173 +32,131 @@ const OIDCConfigurationSchema = z.object({
 
 type OIDCConfiguration = z.infer<typeof OIDCConfigurationSchema>;
 
-const logger = new Logger('OIDCClient');
+@Injectable()
+export class OIDCProvider extends OAuthProvider {
+  override provider = OAuthProviderName.OIDC;
+  #endpoints: OIDCConfiguration | null = null;
 
-class OIDCClient {
-  private static async fetch<T = any>(
-    url: string,
-    options: RequestInit,
-    verifier: z.Schema<T>
-  ): Promise<T> {
-    const response = await fetch(url, options);
+  constructor(private readonly url: URLHelper) {
+    super();
+  }
 
-    if (!response.ok) {
-      logger.error('Failed to fetch OIDC configuration', await response.json());
-      throw new Error(`Failed to configure client`);
+  private get endpoints() {
+    if (!this.#endpoints) {
+      throw new Error('OIDC provider is not configured');
     }
-    const data = await response.json();
-    return verifier.parse(data);
+    return this.#endpoints;
   }
 
-  static async create(config: OAuthOIDCProviderConfig, url: URLHelper) {
-    const { args, clientId, clientSecret, issuer } = config;
-    if (!url.verify(issuer)) {
-      throw new Error('OIDC Issuer is invalid.');
-    }
-    const oidcConfig = await OIDCClient.fetch(
-      `${issuer}/.well-known/openid-configuration`,
-      {
-        method: 'GET',
-        headers: { Accept: 'application/json' },
-      },
-      OIDCConfigurationSchema
-    );
+  override get configured() {
+    return this.#endpoints !== null;
+  }
 
-    return new OIDCClient(clientId, clientSecret, args, oidcConfig, url);
+  protected override setup() {
+    const validate = async () => {
```

Comment:
This `setup` phase is asynchronous, how does the service start detection tell when to wait for it to complete?

Reply:
`super.setup` will check `configured` and add provider to registery